lib/wovnrb.rb

```diff
@@ -66,22 +68,26 @@ def call(env)
 
     def switch_lang(headers, body)
       translated_body = []
+
+      # Must use `.each` for to support multiple-chunks in Sinatra
       string_body = ''
       body.each { |chunk| string_body += chunk }
       html_body = Helpers::NokogumboHelper::parse_html(string_body)
 
       if !wovn_ignored?(html_body) && !amp?(html_body)
-        if html_body.html?
-          # TODO: set lang property to HTML
-          # TODO: insert hreflangs
-          # TODO: insert fallback snippet
-        end
+        html_converter = HtmlConverter.new(string_body, @store, headers)
+        string_body = html_converter.build if html_body.html?
 
         if needs_api?(html_body, headers)
-          # TODO: remove ignored content UNLESS SETTING (SEE WOVN.php)
-          translated_content = ApiTranslator.new(@store, headers).translate(string_body)
-          # TODO: put back ignored content UNLESS SETTING (SEE WOVN.php)
-          translated_body.push(translated_content)
+          converted_html, marker = html_converter.build_api_compatible_html
+          translated_content = ApiTranslator.new(@store, headers).translate(converted_html)
+          reverted_content =
+            if translated_content
```

Comment:
The `ApiTranslator` is already making the choice so `translated_content` is always either a valid translation or the original content, so you can simplify the code and get read of `reverted_content` which is misleading name:
```
converted_html, marker = html_converter.build_api_compatible_html
translated_content = ApiTranslator.new(@store, headers).translate(converted_html)

translated_body.push(marker.revert(translated_content))
```